README.md
# Helix Query Index

> Run queries against an Algolia index on Adobe I/O Runtime

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-index-query.svg)](https://codecov.io/gh/adobe/helix-query-index)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-query-index.svg)](https://circleci.com/gh/adobe/helix-query-index)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-query-index.svg)](https://github.com/adobe/helix-query-index/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-query-index.svg)](https://github.com/adobe/helix-query-index/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-query-index.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-query-index)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release) 

## Installation

## Usage

```bash
curl https://adobeioruntime.net/api/v1/web/helix/helix-services/query-index@v1
```

For more, see the [API documentation](docs/API.md).

## Development

### Deploying Helix Service

Deploying Helix Query Index requires the `wsk` command line client, authenticated to a namespace of your choice. For Project Helix, we use the `helix` namespace.

All commits to main that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/query-index@ci<num>` and tagged with the CI build number.



CHANGELOG.md
## [1.2.122](https://github.com/adobe/helix-query-index/compare/v1.2.121...v1.2.122) (2021-09-03)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.2.3 ([#380](https://github.com/adobe/helix-query-index/issues/380)) ([2879d13](https://github.com/adobe/helix-query-index/commit/2879d13daa6afe4d736aba9978528cd5b405285e))

## [1.2.121](https://github.com/adobe/helix-query-index/compare/v1.2.120...v1.2.121) (2021-08-30)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.2.2 ([#378](https://github.com/adobe/helix-query-index/issues/378)) ([1e51238](https://github.com/adobe/helix-query-index/commit/1e5123842826b7eb399c0f199de2134399c3d808))

## [1.2.120](https://github.com/adobe/helix-query-index/compare/v1.2.119...v1.2.120) (2021-08-21)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.2.1 ([#376](https://github.com/adobe/helix-query-index/issues/376)) ([b729896](https://github.com/adobe/helix-query-index/commit/b729896fd54b5bfabf909d69105681a08d902cd2))

## [1.2.119](https://github.com/adobe/helix-query-index/compare/v1.2.118...v1.2.119) (2021-08-19)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.2.0 ([#374](https://github.com/adobe/helix-query-index/issues/374)) ([a516793](https://github.com/adobe/helix-query-index/commit/a5167934a1aabc97664f1036d4ae447ba2f35195))

## [1.2.118](https://github.com/adobe/helix-query-index/compare/v1.2.117...v1.2.118) (2021-08-18)


### Bug Fixes

* **deps:** update adobe fixes ([#373](https://github.com/adobe/helix-query-index/issues/373)) ([8edeef8](https://github.com/adobe/helix-query-index/commit/8edeef84dcea12c26826f26cad5552559627b369))

## [1.2.117](https://github.com/adobe/helix-query-index/compare/v1.2.116...v1.2.117) (2021-08-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v2.4.2 ([#372](https://github.com/adobe/helix-query-index/issues/372)) ([c4eb3cf](https://github.com/adobe/helix-query-index/commit/c4eb3cf22e09384445de895b5b492c8a1da7418d))

## [1.2.116](https://github.com/adobe/helix-query-index/compare/v1.2.115...v1.2.116) (2021-08-17)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.14 ([#371](https://github.com/adobe/helix-query-index/issues/371)) ([563af08](https://github.com/adobe/helix-query-index/commit/563af08d3a8ba70287a0ed38fcc07364af2352ca))

## [1.2.115](https://github.com/adobe/helix-query-index/compare/v1.2.114...v1.2.115) (2021-08-09)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.13 ([#369](https://github.com/adobe/helix-query-index/issues/369)) ([243106d](https://github.com/adobe/helix-query-index/commit/243106d511f32f4cddc12c6e580cd18114d20000))

## [1.2.114](https://github.com/adobe/helix-query-index/compare/v1.2.113...v1.2.114) (2021-08-01)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.12 ([#363](https://github.com/adobe/helix-query-index/issues/363)) ([9177008](https://github.com/adobe/helix-query-index/commit/917700891369579e1ac1f2bb5a0235f71972c1c5))

## [1.2.113](https://github.com/adobe/helix-query-index/compare/v1.2.112...v1.2.113) (2021-08-01)


### Bug Fixes

* **deps:** update adobe fixes ([#361](https://github.com/adobe/helix-query-index/issues/361)) ([01ca0ca](https://github.com/adobe/helix-query-index/commit/01ca0caa58d12d84f5f46938b963eee30cda2115))

## [1.2.112](https://github.com/adobe/helix-query-index/compare/v1.2.111...v1.2.112) (2021-07-28)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.0 ([#360](https://github.com/adobe/helix-query-index/issues/360)) ([70f1244](https://github.com/adobe/helix-query-index/commit/70f1244e58d9f0434dac73b2b5f1ffb24836d45c))

## [1.2.111](https://github.com/adobe/helix-query-index/compare/v1.2.110...v1.2.111) (2021-07-28)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.11 ([#359](https://github.com/adobe/helix-query-index/issues/359)) ([4e8b762](https://github.com/adobe/helix-query-index/commit/4e8b76248266d60ba36e456367d81db3590bbaa8))

## [1.2.110](https://github.com/adobe/helix-query-index/compare/v1.2.109...v1.2.110) (2021-07-27)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.10 ([#358](https://github.com/adobe/helix-query-index/issues/358)) ([a4dd03c](https://github.com/adobe/helix-query-index/commit/a4dd03c4e5a512c4a4e0925d395a947ef56e73b7))

## [1.2.109](https://github.com/adobe/helix-query-index/compare/v1.2.108...v1.2.109) (2021-07-18)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.9 ([#356](https://github.com/adobe/helix-query-index/issues/356)) ([4d24650](https://github.com/adobe/helix-query-index/commit/4d24650bb67ecc36f607782b7f45a19dce15dee4))

## [1.2.108](https://github.com/adobe/helix-query-index/compare/v1.2.107...v1.2.108) (2021-07-13)


### Bug Fixes

* **deps:** update adobe fixes ([#354](https://github.com/adobe/helix-query-index/issues/354)) ([90fb410](https://github.com/adobe/helix-query-index/commit/90fb410d6676ea6ee223059d62b5952c744558c2))

## [1.2.107](https://github.com/adobe/helix-query-index/compare/v1.2.106...v1.2.107) (2021-07-05)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.7 ([#351](https://github.com/adobe/helix-query-index/issues/351)) ([62d368d](https://github.com/adobe/helix-query-index/commit/62d368dc506f49b962affba85fe3cc245b2045de))

## [1.2.106](https://github.com/adobe/helix-query-index/compare/v1.2.105...v1.2.106) (2021-06-14)


### Bug Fixes

* **deps:** update dependency @adobe/helix-shared-config to v1.1.6 ([#348](https://github.com/adobe/helix-query-index/issues/348)) ([c72cbb2](https://github.com/adobe/helix-query-index/commit/c72cbb2d6acf62d6e7cbe6c222d90f91f5ad26c3))

## [1.2.105](https://github.com/adobe/helix-query-index/compare/v1.2.104...v1.2.105) (2021-06-13)


### Bug Fixes


test/index.test.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
process.env.HELIX_FETCH_FORCE_HTTP1 = true;

const assert = require('assert');
const path = require('path');
const NodeHttpAdapter = require('@pollyjs/adapter-node-http');
const { setupMocha: setupPolly } = require('@pollyjs/core');
const FSPersister = require('@pollyjs/persister-fs');
const index = require('../src/index.js').main;

describe('Index Tests', () => {
  it('index function is present', async () => {
    const result = await index({});
    assert.equal(result.statusCode, 404);
  });

  it('index function returns an object', async () => {
    const result = await index({});
    assert.equal(typeof result, 'object');
  });
});

describe('Helix-Demo Query Generation', () => {
  setupPolly({
    recordIfMissing: false,
    recordFailedRequests: true,
    logging: false,
    adapters: [NodeHttpAdapter],
    persister: FSPersister,
    persisterOptions: {
      fs: {
        recordingsDir: path.resolve(__dirname, 'fixtures/recordings'),
      },
    },
    matchRequestsBy: {
      headers: {
        exclude: ['user-agent'],
      },
    },
  });

  it('non-existing repo looks good', async () => {
    const result = await index({
      __ow_path: '/blog-posts/all',
      __hlx_owner: 'trieloff',
      __hlx_repo: 'helix-foo',
      __hlx_ref: '3aea5fd4cd4d40f5f7c6ce3d74c6f20999903cd3',
    });
    assert.equal(result.statusCode, 404);
  });

  it('non-existing query looks good', async () => {
    const result = await index({
      __ow_path: '/blog-posts/unknown',
      __hlx_owner: 'trieloff',
      __hlx_repo: 'helix-demo',
      __hlx_ref: 'ade6426d0ff01b543bfa06841bb184997924336a',
    });
    assert.equal(result.statusCode, 404);
  });

  it('all query URL looks good', async () => {
    const result = await index({
      __ow_path: '/blog-posts/all',
      __hlx_owner: 'trieloff',
      __hlx_repo: 'helix-demo',
      __hlx_ref: 'ade6426d0ff01b543bfa06841bb184997924336a',
    });

    assert.equal(result.statusCode, 307);
    assert.equal(result.headers['X-Content-Type'], 'application/json');
    assert.equal(result.headers['Cache-Control'], 's-maxage=600');
    assert.equal(result.headers.Location, '/1/indexes/trieloff--helix-demo--blog-posts?query=*&hitsPerPage=25');
  });

  it('by-author query URL looks good', async () => {
    const result = await index({
      __ow_path: '/blog-posts/by-author',
      __hlx_owner: 'trieloff',
      __hlx_repo: 'helix-demo',
      __hlx_ref: 'ade6426d0ff01b543bfa06841bb184997924336a',
      author: 'Lars',
    });

    assert.equal(result.statusCode, 307);
    assert.equal(result.headers['X-Content-Type'], 'application/json');
    assert.equal(result.headers['Cache-Control'], 's-maxage=300');
    assert.equal(result.headers.Location, '/1/indexes/trieloff--helix-demo--blog-posts?query=*&hitsPerPage=25&filters=author%3ALars');
  });

  it('by-author query URL looks good in Sharepoint', async () => {
    const result = await index({
      __ow_path: '/blog-posts/by-author',
      __hlx_owner: 'trieloff',
      __hlx_repo: 'helix-demo',
      __hlx_ref: '09f82a9e4e9a8b5002240aa9fb6e2002f0a9bed7',
      author: 'Lars',
    });

    assert.equal(result.statusCode, 307);
    assert.equal(result.headers['X-Content-Type'], 'application/json');
    assert.equal(result.headers['Cache-Control'], 's-maxage=300');
    assert.equal(result.headers.Location, 'https://adobeioruntime.net/api/v1/web/helix/helix-services/data-embed@v1?hlx_group.p.or=false&hlx_group.1_property.property=author&hlx_group.1_property.value=Lars&src=https%3A%2F%2Fadobe.sharepoint.com%2F%3Ax%3A%2Fr%2Fsites%2FTheBlog%2FShared%2520Documents%2Ftheblog%2Fen%2Fquery-index.xlsx%3Fd%3Dwe7bf6b3af3234076968b30a1565f2373%26csf%3D1%26web%3D1%26e%3Dq9o8tW');
  });
});


test/visit.test.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-param-reassign */

const assert = require('assert');
const visit = require('../src/visitor');

describe('Visitor Unit Tests', () => {
  it('Turns an object uppercase', () => {
    const before = {
      foo: 'bar',
      baz: [
        'zip',
        'zap',
      ],
    };

    const after = {
      foo: 'BAR',
      baz: [
        'ZIP',
        'ZAP',
      ],
    };

    visit(before, (key, value, obj) => {
      if (typeof value === 'string') {
        obj[key] = value.toUpperCase();
      }
    });

    assert.deepEqual(before, after);
  });

  it('Turns an array uppercase', () => {
    const before = [{
      foo: 'bar',
      baz: [
        'zip',
        'zap',
      ],
    }];

    const after = [{
      foo: 'BAR',
      baz: [
        'ZIP',
        'ZAP',
      ],
    }];

    visit(before, (key, value, obj) => {
      if (typeof value === 'string') {
        obj[key] = value.toUpperCase();
      }
    });

    assert.deepEqual(before, after);
  });
});


test/post-deploy.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */

const assert = require('assert');
const chai = require('chai');
const chaiHttp = require('chai-http');
const packjson = require('../package.json');

chai.use(chaiHttp);
const { expect } = chai;

function getbaseurl() {
  const namespace = 'helix';
  const package = 'helix-services';
  const name = packjson.name.replace('@adobe/helix-', '');
  let version = `${packjson.version}`;
  if (process.env.CI && process.env.CIRCLE_BUILD_NUM && process.env.CIRCLE_BRANCH !== 'main') {
    version = `ci${process.env.CIRCLE_BUILD_NUM}`;
  }
  return `api/v1/web/${namespace}/${package}/${name}@${version}`;
}

describe('Post-Deploy Tests', () => {
  it('Service is ready for monitoring', () => {
    assert.equal(
      'I am ready to go on call for this',
      'I am ready to go on call for this',
    );
  });

  it('All Blog Posts', async () => {
    const path = '/blog-posts/all?__hlx_owner=trieloff&__hlx_repo=helix-demo&__hlx_ref=ade6426d0ff01b543bfa06841bb184997924336a';

    await chai
      .request('https://adobeioruntime.net/')
      .get(`${getbaseurl()}${path}`)
      .redirects(0)
      .then((response) => {
        expect(response).to.redirect;
        expect(response).to.redirectTo('/1/indexes/trieloff--helix-demo--blog-posts?query=*&hitsPerPage=25');
      })
      .catch((e) => {
        throw e;
      });
  }).timeout(10000);
});


src/visitor.js
/*
 * Copyright 2020 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

function visit(root, callback) {
  if (Array.isArray(root)) {
    root.forEach((value, index) => {
      callback(index, value, root);
      visit(value, callback);
    });
  } else if (typeof root === 'object') {
    Object.entries(root).forEach(([key, value]) => {
      callback(key, value, root);
      visit(value, callback);
    });
  }
}

module.exports = visit;


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const wrap = require('@adobe/helix-shared-wrap');
const { logger } = require('@adobe/openwhisk-action-logger');
const { wrap: status } = require('@adobe/helix-status');
const { epsagon } = require('@adobe/helix-epsagon');
const { qb } = require('@adobe/helix-querybuilder');
const fetchAPI = require('@adobe/helix-fetch');
const { IndexConfig } = require('@adobe/helix-shared-config');

const { fetch } = process.env.HELIX_FETCH_FORCE_HTTP1
  /* istanbul ignore next */
  ? fetchAPI.context({
    userAgent: 'helix-fetch', // static user-agent for recorded tests
    alpnProtocols: [fetchAPI.ALPN_HTTP1_1],
  })
  /* istanbul ignore next */
  : fetchAPI;

function getIndex(i, name) {
  const [myindex] = i.indices.filter((idx) => idx.name === name);
  return myindex;
}

function getURL(target, query, paramlist, params) {
  const embedurl = new URL('https://adobeioruntime.net/api/v1/web/helix/helix-services/data-embed@v1');

  const searchparams = qb.url(query, 'hlx_');
  searchparams.forEach((value, name) => {
    searchparams.set(name, IndexConfig.evaluate(value, paramlist, params));
  });

  searchparams.append('src', target);

  embedurl.search = searchparams;

  return embedurl.href;
}

/**
 * This is the main function
 * @param {string} name name of the person to greet
 * @returns {object} a greeting
 */
async function main(params) {
  const { __ow_logger: log } = params;
  const cleanparams = Object.entries(params).reduce((clean, [key, value]) => {
    if (!key.startsWith('__')) {
      // eslint-disable-next-line no-param-reassign
      clean[key] = value;
    }
    return clean;
  }, {});

  const path = params.__ow_path || '';
  const [, index, query] = path.split('/');
  /* eslint-disable no-underscore-dangle */
  const owner = params.__hlx_owner;
  const repo = params.__hlx_repo;
  const ref = params.__hlx_ref;
  /* eslint-enable no-underscore-dangle */

  if (!(index && query && owner && repo && ref)) {
    log.warn(`Missing parameters for ${index}/${query} and ${owner}/${repo}/${ref}`);
    return {
      statusCode: 404,
      body: 'Invalid index or query or missing owner, repo, and ref.',
      headers: {
        'Cache-Control': 's-maxage=600',
      },
    };
  }

  const resp = await fetch(`https://raw.githubusercontent.com/${owner}/${repo}/${ref}/helix-query.yaml`);
  if (resp.status !== 200) {
    log.warn(`Missing index config for ${owner}/${repo}/${ref}`);
    return {
      statusCode: 404,
      body: 'Index configuration does not exist',
      headers: {
        'Cache-Control': 's-maxage=600',
      },
    };
  }

  const yamltext = await resp.text();
  const config = await new IndexConfig().withSource(yamltext).init();

  const i = getIndex(config, index);
  const q = config.getQuery(index, query);
  if (q
    && typeof q.query !== 'string'
    && i.target
    && (i.target.match(/^https:\/\/docs\.google\.com\/spreadsheets\/d\/.*/)
    || i.target.match(/^https:\/\/.*\.sharepoint\.com\//))) {
    return {
      statusCode: 307,
      headers: {
        Location: getURL(i.target, q.query, q.parameters, cleanparams),
        'X-Content-Type': 'application/json',
        'X-Static': 'Raw/Query',
        'Cache-Control': `s-maxage=${config.getQueryCache(index, query)}`,
      },
    };
  }

  const location = config.getQueryURL(index, query, owner, repo, cleanparams);
  if (!location) {
    log.warn(`Unknown query ${index}/${query} for ${owner}/${repo}/${ref}`);
    return {
      statusCode: 404,
      body: 'Specified index not found in index configuration',
      headers: {
        'Cache-Control': 's-maxage=600',
      },
    };
  }

  return {
    statusCode: 307,
    headers: {
      // indexname, queryname, owner, repo, urlparams)
      Location: config.getQueryURL(index, query, owner, repo, cleanparams),
      'X-Content-Type': 'application/json',
      'X-Static': 'Raw/Query',
      'Cache-Control': `s-maxage=${config.getQueryCache(index, query)}`,
    },
  };
}

module.exports.main = wrap(main)
  .with(epsagon)
  .with(status)
  .with(logger.trace)
  .with(logger);
